Review PR for Matteo

Fendross opened the pull request:
Ti lascio gli step che ho seguito:
- cancellato e creato un nuovo repo;
- inizializzato su un altro folder locale e committato il readme (probabilmente questo passaggio poteva essere evitato? ho visto che mi richiedeva la creazione di un file prima di fare una PR);
- creato il branch review-matteo e committato tutti i file;
- ti ho aggiunto come collaboratore e ti sto ora mandando la richiesta come reviewer;

Fammi sapere (senza fretta, ovviamente!) se è tutto corretto come procedimento.
Grazie in anticipo!

bigliolimatteo commented:
Hey there 😃 

I'll switch to ENG cause it's kinda the default language for this kind of environment 😂 

The steps I think are correct 👍  I would have done the same thing (maybe just select a better name for the branch next time, something like "first-implementation" but nothing to worry about 😃 )!

I'll check the code ASAP and I warn you I'll be as mean as possible in order to maximise your learning 🧠 😂

bigliolimatteo commented on requirements.txt:
```diff
@@ -0,0 +1,2 @@
+pygame
```
Note that it is usually better to specify the version as well in the requirements, something like `pygame==2.1.2`

It is not mandatory (I don't do that every time 😂 ) but it can help to avoid dependency problems

if you are using a [venv](https://docs.python.org/3/library/venv.html) in python you can easily compute the requirements.txt file with [pip freeze](https://pip.pypa.io/en/stable/cli/pip_freeze/)

Fendross replied:
Ok perfect! I actually wasn't sure about which one to add. In situations where (I feel like) my code should work on at least the latest version, should I just add that? Otherwise, is it the best solution to test previous versions (up to when it makes sense to check them) and go with the first one that gives some problems?

bigliolimatteo replied:
You dont need to test previous version, usually you should work with the latest version (apart from situations in which you need an older version) and specify that in the requirements!

Ideally in a well maintained codebase you will release new versions of your code updated with new versions of your dependencies 😃

bigliolimatteo commented on main.py:
```diff
@@ -0,0 +1,118 @@
+    pg.display.set_caption('Game of Life')
+
+    # Initialises the cells 2D array, fills the screen with the GRID_COLOR and shows the initial state of the simulation
+    cells = np.zeros((6*CELL_SIZE, 8*CELL_SIZE))
```
Why `6` and `8`? Are these number related to the `80` and `60` of line 74?

It is usually best to avoid hardcoding constants like this in the code without a proper explanation, a better way to do this could be to add them to the list of constants that you have at the beginning of the file, sth like `VERTICAL_BLABLA=6` and `HORIZONTAL_BLABLA=8`

Fendross replied:
Yes that is a hardcoded solution, I couldn't work around the size of the grid properly and ended up with this. I will refactor this!

bigliolimatteo replied:
🚀

bigliolimatteo commented on main.py:
```diff
@@ -0,0 +1,118 @@
+    pg.display.flip()
+    pg.display.update()
+
+    is_running = False
```
This `is_running ` concept is hard to understand, is there a way (more comments or different code) so that it can be more understandable?

Fendross replied:
Ok! I will think of a way to refactor this, if nothing better comes to mind I will comment it properly.

bigliolimatteo commented on main.py:
```diff
@@ -0,0 +1,118 @@
+    # Iterating over every row and col of cells
+    for row, col in np.ndindex(cells.shape):
+        # Calculate the number of alive neighbor cells and subtract the contribution of the current cell, it doesn't count as a neighbor
+        alive = np.sum(cells[row - 1 : row + 2, col - 1 : col + 2]) - cells[row, col]
```
`alive` is quite generic as a name, usually it is better to use longer names in order to make stuff more easy to read/track also for people who see just a piece of the code (something like `number_of_alive_neighbours`)

Fendross replied:
✅

bigliolimatteo commented on main.py:
```diff
@@ -0,0 +1,118 @@
+        alive = np.sum(cells[row - 1 : row + 2, col - 1 : col + 2]) - cells[row, col]
+
+        # Get the color of the current cell
+        if cells[row, col] == 0:
```
NOT a problem but know that you can write this also as `color = BG_COLOR if cells[row, col] == 0 else ALIVE_CELL_COLOR `

Fendross replied:
Ok! A more condensed syntax!

bigliolimatteo commented on main.py:
```diff
@@ -0,0 +1,118 @@
+        else:
+            color = ALIVE_CELL_COLOR
+
+        # Game of Life rules
```
Maybe it's better to split the `with_progress` and non-`with_progress ` in two separate blocks, otherwise it's hard to comprehend what is going on

(you could even create two different functions, one to just "reset" the screen and one to actually "play")

Fendross replied:
Reading it back it's kinda messy, I agree. Will rewrite this as well!

bigliolimatteo commented:
+ It is usually better to add a `.gitignore` file to the whole repo in order to avoid uploading useless file (e.g. the `.DS_Store` file should not be here)

Fendross commented:
Good afternoon!

After your review I decided to tackle most of the issues you brought forward. 

## Timeline

### [c7b5d31](https://github.com/Fendross/game-of-life/pull/1/commits/c7b5d31b7644c11ad38d0724b84a56dda7e788a9) commit: 
- Refactored the application of rules in a new function, `apply_rules`, to thin out the `update` function;
- Parameterized most of the code;
- Gave less generic names (e.g., `is_simulation_active`, `number_of_alive_neighbors`) to variables;
- Added more comments (maybe too many? I wanted to ask you this since I found myself writing almost one for each line of code).

### [33d8c10](https://github.com/Fendross/game-of-life/pull/1/commits/33d8c10c1c107c43f40a830662cedfdea4fd77ae) commit:
- Added a newline to the **.gitignore** file, since it was complaining.

Thank you in advance!

bigliolimatteo commented:
General comments:

- Yeah maybe there are now too many long comments 😂  but that's kinda of an art to be able to write less and small comments but still useful and reliable, dont worry!
-  `**/.DS_Store` should be just `.DS_Store`, and because it is in the gitignore, it should not be present in the PR 😱

bigliolimatteo commented on .gitignore:
```diff
@@ -0,0 +1 @@
+**/.DS_Store
```
look at general comment
